gcovr/gcov.py

```diff
@@ -298,25 +300,40 @@ def parse_line(
         if self.excluding and not self.excluding[-1]:
             self.excluding.pop()
 
-    def parse_code_line(self, status, code):
+    def parse_code_line(self, status, code, exclude_function_coverage):
         firstchar = status[0]
 
+        last_was_function_marker = self.last_was_function_marker
+        self.last_was_function_marker = False
+
         if firstchar == '-' or (self.excluding and firstchar in "#=0123456789"):
             # remember certain non-executed lines
             if self.excluding or is_non_code(code):
+                if last_was_function_marker:
+                    if exclude_function_coverage:
+                        return True
+                    self.coverage.line(self.lineno).function = True
                 self.coverage.line(self.lineno).noncode = True
             return True
 
         # "#": uncovered
         # "=": uncovered, but only reachable through exceptions
         if firstchar in "#=":
+            if last_was_function_marker:
+                if exclude_function_coverage:
+                    return True
+                self.coverage.line(self.lineno).function = True
             if self.excluding or is_non_code(code):
                 self.coverage.line(self.lineno).noncode = True
```

Comment:
I'Ve tried this, but also the counter and and noncode has to be set outside because if the line is excluded it must not be added in the if/elif construct. I set the noncode also if the function coverage isn't excluuded.
I'm not sure if this is correct.

Reply:
Ok, that looks better but it the code seems incorrect – all tests were updated, which can't be the case (?). Apparently now all function lines are being excluded by default?

I think the problem is this snippet:

```python
if not (last_was_function_marker and exclude_function_lines):
    if last_was_function_marker:
        noncode = True
```

We now have `last_was_function_marker = True` and `exclude_function_lines = False`. The outer condition will evaluate to True, but the inner condition will set `noncode = True` nevertheless.

I think that section should be rewritten like this:

```python
if exclude_function_coverage and last_was_function_marker:
    noncode = True

if noncode:
    self.coverage.line(self.lineno).noncode = True
if count is not None:
    self.coverage.line(self.lineno).count += count
```

Or maybe this `noncode = True` line should be a `return True`?

Reply:
That's why I wasn't sure. If we set the line flag `noncode = True` we shouldn't increment the line counter. In my opionion a noncode line can't have a counter not zero. I'll check your suggestions, this looks better. I'll think it should be:
```python
if noncode:
    self.coverage.line(self.lineno).noncode = True
elif count is not None:
    self.coverage.line(self.lineno).count += count```